TESTS/mbed_hal/rtc/main.cpp

```diff
@@ -26,34 +26,96 @@
 #include "mbed.h"
 #include "rtc_api.h"
 
-
 using namespace utest::v1;
 
 static const uint32_t WAIT_TIME = 4;
 static const uint32_t WAIT_TOLERANCE = 1;
 
+#define US_PER_SEC 1000000
+#define ACCURACY_FACTOR 10
 
-void rtc_init_test()
+static const uint32_t DELAY_4S = 4;
+static const uint32_t DELAY_10S = 10;
+static const uint32_t RTC_TOLERANCE = 1;
+static const uint32_t TOLERANCE_ACCURACY_US = (DELAY_10S * US_PER_SEC / ACCURACY_FACTOR);
+
+#if DEVICE_LOWPOWERTIMER
+volatile bool expired;
+
+void callback(void)
 {
-    for (int i = 0; i < 10; i++) {
-        rtc_init();
-    }
+    expired = true;
 }
 
-void rtc_sleep_test()
+/* Auxiliary function to test if RTC continue counting in
+ * sleep and deep-sleep modes. */
+void rtc_sleep_test_support (bool deepsleep_mode)
 {
+    LowPowerTimeout timeout;
     const uint32_t start = 100;
+    expired = false;
+
+    /*
+     * Since deepsleep() may shut down the UART peripheral, we wait for 10ms
+     * to allow for hardware serial buffers to completely flush.
+     * This should be replaced with a better function that checks if the
+     * hardware buffers are empty. However, such an API does not exist now,
+     * so we'll use the wait_ms() function for now.
+     */
+    wait_ms(10);
+
     rtc_init();
 
+    if(deepsleep_mode == false) {
+        sleep_manager_lock_deep_sleep();
+    }
+
     rtc_write(start);
-    wait(WAIT_TIME);
+
+    timeout.attach(callback, DELAY_4S);
+
+    TEST_ASSERT(sleep_manager_can_deep_sleep() == deepsleep_mode);
+
+    while(!expired) sleep();
+
     const uint32_t stop = rtc_read();
 
+    TEST_ASSERT_UINT32_WITHIN(RTC_TOLERANCE, DELAY_4S, stop - start);
+
+    timeout.detach();
+
+    if(deepsleep_mode == false) {
+        sleep_manager_unlock_deep_sleep();
+    }
+
     rtc_free();
+}
+#endif
 
-    TEST_ASSERT_UINT32_WITHIN(WAIT_TOLERANCE, WAIT_TIME, stop - start);
+/* Test that ::rtc_init can be called multiple times. */
+void rtc_init_test()
+{
+    for (int i = 0; i < 10; i++) {
+        rtc_init();
```

Comment:
I think we could check if it actually continue working.

Reply:
This test case only checks if init can be called multiple times. rtc_persist_test() verifies if RTC still counts after free/init.